release: 2.76 (#17117)
snapcore/snapd: NEWS.md

* LP: #2110368 Fix download of private snap components by setting UserID
* LP: #2144666 Fix mount namespace updates with synthetic bind mounts on same target paths
* LP: #2146337 Improve handling of failed downloads and retain partial files for resume
* LP: #2147207 Fix snap enable/disable cycle forgetting components
* Make run-inhibit hint for kill-snap-apps task based on kill reason
* Merge content-provider prerequisite updates into seed-refresh
* Move `SortServices` into `Backend.StartServices`
* Move state to client change conversion to ctlcmd package
* Omit misleading "try to refresh snapd" suggestion for ISA-related errors
* Only create `link-component` tasks when needed during refresh to existing revision
* Reconfigure piboot bootloader on gadget refreshes to preserve `os_prefix`
* Reduce the number of AppArmor profile regenerations during snap operations
* Refactor seed-refresh ownership to `devicestate`
* Regenerate certificate database on remodels
* Remove obsolete FIXME comment in `VersionCompare`
* Remove unused `GenerateDmVerityData` helper from snap/integrity
* Rename and document error type for ISA assumes flags

> * Replace `rootPath` variable with `dirs.GlobalRootDir` in sandbox packages

Reviewer: This is an internal refactor, not really worth mentioning, but it's fine

```suggestion
```

Author: fixed

* snap: allow removing a snap and its base at the same time
* snap: display detailed component information in `snap info`
* snap: extend `AlreadyInstalledError` to multiple snaps and components
* snap: extend `set-quota` command options description with accepted value formats
* snap: implement `snap delta` command for computing snap deltas
* snap: improve consistency for `snap install` when some snaps are already installed
* snap: show hint in `snap list` that a snap has components
* snap-confine: allow linking to `libm` in AppArmor profile
* snap-confine: fix out-of-bounds read in mountinfo parser for partial escape sequences
* snap-confine: harden bpffs mount with `nosuid`, `nodev`, `noexec` flags
* snap-confine: remove experimental persistent per-user mount namespace feature
* snap-confine: set `FD_CLOEXEC` on file descriptors returned by BPF helpers
* snap-confine: support `transparent_hugepage` in AppArmor profile
* snap-confine: use `strchr` after NUL-terminating in infofile parser
* snap-update-ns: switch to a multi-pass process for constructing and updating mount namespaces
* `RemoveMountUnitFile` now unmounts even if mount unit file is missing

> * Add `dlv` support to `IsTestBinary` function

Author: fixed

* Add explicit mount phase during single-reboot refresh to fix undo of kernel refreshes
* Add security audit logging subsystem
* Allow `openshell` snap to use experimental `daemon-scope: user`
* Allow configuring mount unit options based on filesystem type
* Allow equals signs in uevent values in netlink parser
* Also bind-mount directories modified by kmod backend during preseed
* Clean up potentially corrupted files during snap download undo
* Complete the bootloader environment implementation
* Copy integrity data files during snap install
* Create hook for seed refresh mode
* Create removal tasks for old seed-refresh seeds
* Dispatch `systemctl` commands asynchronously when calling `Stop()`
* Ensure `/tmp/.X11-unix` created inside mount namespace has correct permissions
* Ensure exclusive changes conflict with refresh/revert
* Export ubuntu-boot-state filename constant from bootloader package
* Fix duplicate removal of apps under `$SNAP_MOUNT_DIR/bin`

> * Fix govet errors by inlining deprecated `reflect.Ptr` constant

Author: removed

* Fix integration between prerequisites task and seed-refresh mode
* Fix split-refresh overwriting provided lane
* Fix use of umask in `GetListener` for socket activation
* Ignore `net.ErrClosed` during daemon shutdown
* Implement `ResolveValidationSetsEnforcementError` in terms of one call
* Improve `snapctl install` consistency when components are already installed

> * Improve dot graph test integration for state debugging

Author: removed

* Merge content-provider prerequisite updates into seed-refresh
* Move `SortServices` into `Backend.StartServices`
* Move state to client change conversion to ctlcmd package
* Omit misleading "try to refresh snapd" suggestion for ISA-related errors
* Only create `link-component` tasks when needed during refresh to existing revision
* Reconfigure piboot bootloader on gadget refreshes to preserve `os_prefix`
* Reduce the number of AppArmor profile regenerations during snap operations
* Refactor seed-refresh ownership to `devicestate`
* Regenerate certificate database on remodels
* Remove obsolete FIXME comment in `VersionCompare`
* Remove unused `GenerateDmVerityData` helper from snap/integrity
* Rename and document error type for ISA assumes flags
* Replace `rootPath` variable with `dirs.GlobalRootDir` in sandbox packages
* Restart snapd from `daemon.Stop` to improve restart reliability
* Restart stopped services on error in `stopSnapServices` for transactionality
* Simplify certificate-db updates on model-base refresh/installs
* Support racing `Loop` and `Stop` correctly in overlord
* Support sending file descriptors to systemd via `sd_notify`

> * Transition to using spread-plus for integration testing

Author: removed

# New in snapd 2.76
* assertions: add helper for validating integrity data
* assertions: drop incorrect/non-standard Ed25519 support

> * confdb: allow API admin access to confdb secrets

Reviewer 2: ```suggestion
* confdb: allow only API admin read access to confdb secrets
```

Author: fixed

* confdb: block concurrent confdb accesses
* confdb: block concurrent snapctl accesses to configuration database
* confdb: check for ephemeral data when missing `save-view` hook on commit
* confdb: ignore not-found errors in confdb-schema refreshes

> * confdb: include user ID when getting confdb values

Reviewer 2: ```suggestion
```
This is internal-only and part of the general secrets implementation

Author: fixed

* confdb: support `--wait-for` timeouts when accessing confdb
* core-initrd: add group referenced in udev rules
* core-initrd: add libbpf dependency to initramfs
* core-initrd: add missing libbpf dependency in 24.04 packaging
* core-initrd: ensure audio is a system group
* core-initrd: fix `/boot/uboot` mount with u-boot env in dedicated partition
* core-initrd: increase mount burst from 5 to 128 for faster boot
* core-initrd: sync partition udev rules with the ones in core-base
* core-initrd: sync with latest upload to snappy-dev PPA
* core-initrd: synchronize changelogs with latest PPA upload
* core-initrd: update changelog with latest PPA upload
* LP: #2150773 core-initrd: add nfnetlink module to fix nf netlink socket speed regression (Ubuntu Core only)
* cross-distro: allow snapd to manipulate systemd unit files in SELinux policy
* cross-distro: FIPS bootstrap and dispatch via `snap-fips-dispatch`
* Update seccomp syscalls list for kernel 7.1.0
* desktop: fix common ID selection with multiple desktop plugs
* FDE: allow user mode on core in secboot TPM handling
* FDE: bump go-efilib dependency

> * FDE: bump secboot for FDE fixes

Reviewer 2: Should this include the secboot commit hash?

Author: fixed

* FDE: deprecate `check-pin`/`passphrase` API endpoints
* FDE: improve tracing for OP-TEE probing
* FDE: move auto-repair logic to `overlord/fdestate` and provide state
* FDE: update secboot for TPM/FDE bug fixes including Intel HAP and recovery key parsing
* FDE: use any primary key matching digest when adding a keyslot
* FDE: use ignore action for preinstall check in VM
* interfaces: bluez | drop explicit `deny send_destination` in D-Bus configuration
* interfaces: conditionally deny `/proc/self/mountinfo` to suppress Go 1.25+ denials
* interfaces: custom-device | fix `for-device` validation panic on non-string value
* interfaces: disallow auto-connect to parallel installs
* interfaces: docker | make plug implicit on classic systems
* interfaces: ignore errors in disconnect hooks during explicit snap disconnect

> * interfaces: log when no effects are scheduled during interface state handling

Reviewer 2: If this is from this PR https://github.com/canonical/snapd/pull/16785, then it should be dropped since it was released in 2.75.2

Author: fixed

* interfaces: mediatek-accel | add plug interface base declaration
* interfaces: microceph-support | suppress noisy sudo denial audit logs

> * interfaces: network-setup-control, network-setup-observe | allow running python in UC26

Reviewer 2: ```suggestion
```
this was in 2.75.2

Author: fixed

* interfaces: podman | add new interface for podman socket access
* interfaces: pulseaudio | fix security tag syntax inconsistency
* interfaces: raw-usb | allow USB device enumeration on Fairphone 5 with NexDock
* interfaces: restore auto-connections on failed refresh undo
* LP: #2148544 interfaces: bool-file | support deep SoC sysfs paths for LED brightness
* LP: #2139213 packaging: make Ubuntu 16.04 packaging dep17 compliant
* packaging: add cross-distro build script and instructions
* packaging: default openSUSE to `/var/lib/snapd/snap` and sync from downstream

> * packaging: depend on Go 1.23 for FIPS builds

Reviewer: Technically this was already the case, this just cleaned up the dependencies so extra go versions weren't installed in parallel. Probably worth removing this entry to avoid confusion.
```suggestion
```

Author: removed

* packaging: drop transitional packages only for Ubuntu 26.04 (Resolute)

> * packaging: fix Debian builds

Reviewer: Were debian builds broken during the last release? If this is just a fix for something which broke since last release, probably shouldn't include this here
```suggestion
```

Author: changed to `packaging: Debian build improvements`. See: https://github.com/canonical/snapd/pull/16704

fixed

* packaging: fix Launchpad FIPS build detection for `snapd-fips` job
* packaging: refactor and clean up `snapd.mk`, standardize test-data directories

> * packaging: revert `mkversion.sh` correct for deb release flow

Reviewer 2: I agree with the bot on this one. I'm not sure what the entry is saying.

> * packaging: revert `mkversion.sh` correct for deb release flow

Author: fixed

* packaging: switch to `golang-github-chai2010-gettext-go-dev`

> * packaging: sync openSUSE packaging from downstream

Reviewer 3: there already is a note about it a bit earlier
```suggestion
```

Author: `packaging: default openSUSE to `/var/lib/snapd/snap` and sync from downstream` -> https://github.com/canonical/snapd/pull/16943 -> https://warthogs.atlassian.net/browse/SNAPDENG-36210

`packaging: sync openSUSE packaging from downstream` -> https://github.com/canonical/snapd/pull/16727 -> https://warthogs.atlassian.net/browse/SNAPDENG-36168

The descriptions sounds related, but its not obvious by looking at PRs and JIRA. 

Anyways, will drop this one, fixed.

* packaging: update spread installation for autopkgtests to use canonical/spread

> * packaging: update to AppArmor 4.1.7

Reviewer 3: this is only about the bundled apparmor, but not sure how to indicate that nicely
```suggestion
* packaging(snap): update bundled to AppArmor 4.1.7
```

Reviewer 4: I agree

Reviewer: My suggestion: "update bundled AppArmor to 4.1.7"

Author: `packaging: update bundled AppArmor 4.1.7 (snapd snap only)` - fixed

> * prompting: clean up mocked and exported functions

Reviewer: Not quite sure what this is, can probably be removed
```suggestion
```

Author: removed

* prompting: escape paths in prompt constraints
* prompting: improve API error handling and validation
* prompting: improve error message when no handler service is present
* prompting: re-enable the prompting notice backend

> * prompting: respond with full permission set user allowed

Reviewer: Sure, though not a big deal

> * prompting: respond with full permission set user allowed

Author: fixed - `prompting: respond with full user-allowed permission set`

* prompting: validate permissions while unmarshalling
* remote device management: implement `dispatch-mgmt-messages` task with sequencing support
* LP: #2125344 snap: avoid empty channel forwarding message
* LP: #2150683 snap: clarify `snap install` help text for `--classic` and `--devmode`
* LP: #2152908 snap: print complex attributes in `snap interface --attrs` output
* snap: add run-inhibit hint and inhibit info when a snap is disabled
* snap: allow removing a snap and its base at the same time
* snap: display detailed component information in `snap info`
* snap: extend `AlreadyInstalledError` to multiple snaps and components
* snap: extend `set-quota` command options description with accepted value formats
* snap: implement `snap delta` command for computing snap deltas
* snap: improve consistency for `snap install` when some snaps are already installed

> * snap: remove incomplete async `snapctl` functionality prior to 2.76

Reviewer 5: Probably would omit this

Reviewer: Agreed

```suggestion
```

Author: fixed

* snap: show hint in `snap list` that a snap has components
* snap-confine: allow linking to `libm` in AppArmor profile
* snap-confine: fix out-of-bounds read in mountinfo parser for partial escape sequences
* snap-confine: harden bpffs mount with `nosuid`, `nodev`, `noexec` flags
* snap-confine: remove experimental persistent per-user mount namespace feature
* snap-confine: set `FD_CLOEXEC` on file descriptors returned by BPF helpers
* snap-confine: support `transparent_hugepage` in AppArmor profile

> * snap-confine: switch `snap-update-ns` to a multi-pass process

Reviewer 3: ```suggestion
* snap-update-ns: switch to a multi-pass process for constructing and updating mount namespaces
```
and should be moved right after the snap-confine block

Author: fixed

* snap-confine: use `strchr` after NUL-terminating in infofile parser
* `RemoveMountUnitFile` now unmounts even if mount unit file is missing

> * Add `--no-wait` flag to `snapctl install` and `snapctl remove`

Reviewer 6: This should be removed as we disabled the async feature

Author: fixed

> * Add `--no-wait` flag to `snapctl install` and `snapctl remove`

Reviewer 5: Same here

Author: fixed

* Add `dlv` support to `IsTestBinary` function

> * Add `snapctl is-ready` sub-command with per-change rate limiting

Reviewer 6: same here

Author: fixed

> * Add `snapctl is-ready` sub-command with per-change rate limiting

Reviewer 5: This should be removed, not in this release

Author: fixed

* Add explicit mount phase during single-reboot refresh to fix undo of kernel refreshes
* Add security audit logging subsystem with nop implementation (part 1)
* Add security audit logging subsystem with slog implementation and audit sink

> * Add security logging for snapd login events via audit subsystem

Reviewer 3: ```suggestion
* Add security audit logging subsystem
```

Author: fixed

> * Allow `openshell` snap to use `daemon-scope: user`

Reviewer 7: @zyga do we want this here?  if we do it should say: experimental `daemon-scope ...`

Author: Changed to 

```Allow `openshell` snap to use experimental `daemon-scope: user```

* Allow configuring mount unit options based on filesystem type
* Allow equals signs in uevent values in netlink parser

> * Allow SELinux mounting on `/var/snap/<snap>/<rev>`

Reviewer 2: ```suggestion
```
this was in 2.75.2

Author: fixed

* Also bind-mount directories modified by kmod backend during preseed
* Clean up potentially corrupted files during snap download undo
* Complete the bootloader environment implementation

> * Consider all tasks in a lane that might be waiting for a reboot

Reviewer 2: ```suggestion
```
this was in 2.75.2

Author: fixed

* Copy integrity data files during snap install
* Create hook for seed refresh mode
* Create removal tasks for old seed-refresh seeds
* Dispatch `systemctl` commands asynchronously when calling `Stop()`
* Ensure `/tmp/.X11-unix` created inside mount namespace has correct permissions
* Ensure exclusive changes conflict with refresh/revert
* Export ubuntu-boot-state filename constant from bootloader package
* Fix duplicate removal of apps under `$SNAP_MOUNT_DIR/bin`
* Fix govet errors by inlining deprecated `reflect.Ptr` constant
* Fix integration between prerequisites task and seed-refresh mode
* Fix split-refresh overwriting provided lane
* Fix use of umask in `GetListener` for socket activation
* Ignore `net.ErrClosed` during daemon shutdown
* Implement `ResolveValidationSetsEnforcementError` in terms of one call
* Improve `snapctl install` consistency when components are already installed
* Improve dot graph test integration for state debugging
* Inject seed creation tasks into snap refresh flow
* Introduce system options for custom certificates on Ubuntu Core
* Keep idle services with activation units stopped on reload
* List snap components in `snap-debug-info` via debug-tools

> * Log skipping boot-ok check only once

Reviewer 2: ```suggestion
```
this was in 2.75.2

Author: fixed

* Look at gadget.yaml instead of marker file to determine ubootpart usage
* LP: #1966067 Skip redundant xdg-settings confirmation prompt when setting is already correct
* LP: #2110368 Fix component installation for private snaps via `snapctl`
* LP: #2110368 Fix download of private snap components by setting UserID
* LP: #2144666 Fix mount namespace updates with synthetic bind mounts on same target paths
* LP: #2146337 Improve handling of failed downloads and retain partial files for resume
* LP: #2147207 Fix snap enable/disable cycle forgetting components
* Make run-inhibit hint for kill-snap-apps task based on kill reason
* Merge content-provider prerequisite updates into seed-refresh
* Move `SortServices` into `Backend.StartServices`
* Move state to client change conversion to ctlcmd package
* Omit misleading "try to refresh snapd" suggestion for ISA-related errors
* Only create `link-component` tasks when needed during refresh to existing revision
* Reconfigure piboot bootloader on gadget refreshes to preserve `os_prefix`

> * Reduce the number of AppArmor profile regenerations during snap operations

Reviewer: Is this really new in this release?

Author: Yes: https://github.com/canonical/snapd/pull/16486

* Refactor seed-refresh ownership to `devicestate`
* Regenerate certificate database on remodels
* Remove obsolete FIXME comment in `VersionCompare`
* Remove unused `GenerateDmVerityData` helper from snap/integrity
* Remove unused argument from `webify` function in daemon

> * Remove unused daemon router calls and rename `webify` to `injectSnapIconURL`

Reviewer: These two could probably be consolidated into one or perhaps removed entirely

Author: dropped both

* Rename and document error type for ISA assumes flags
* Replace `rootPath` variable with `dirs.GlobalRootDir` in sandbox packages
* Restart snapd from `daemon.Stop` to improve restart reliability
* Restart stopped services on error in `stopSnapServices` for transactionality
* Simplify certificate-db updates on model-base refresh/installs
* Support racing `Loop` and `Stop` correctly in overlord
* Support sending file descriptors to systemd via `sd_notify`
* Transition to using spread-plus for integration testing
* Unroll CPU-heavy recursive function in snap state handlers
* Use change ID to prevent nested seed-refresh spawned by prerequisites

> * Use more precise prune pattern for tmpfiles (CVE-2026-3888)

Reviewer 2: ```suggestion
```
this was in 2.75.2

Author: fixed

> * Use user ID when refreshing multiple snaps via the API

Reviewer 2: ```suggestion
```
this was in 2.75.2

Author: fixed
